wallet: Fix migration of wallets with txs that have both spendable and watchonly outputs

achow101 opened the pull request:
A transaction does not necessarily have to belong to either the migrated wallet (with the private keys) and the watchonly wallet (with watchonly things), it could have multiple outputs with each isminetype. So we should be putting such transactions in one or the other wallet, but rather putting it in both.

I've added a test for this behavior, however the test also revealed a few other issues. Notably, it revealed that `migratewallet` would have the watchonly wallet rescan from genesis when it is reloaded at the end of migration. This could be a cause for migration appearing to be very slow. This is resolved by first writing best block records to the watchonly and solvable wallets, as well as updating the test to make sure that rescans don't happen.

The change to avoid rescans also found an issue where some of our early exits would result in unloading the wallet even though nothing happened. So there is also a commit to reload the wallet for such early exits.

achow101 commented:
It seems like there's a random failure happening where something in BDB is `free()`ing an invalid pointer which causes bitcoind to abort. This appears to only be happening with the change the reloads the wallet for the early exits. I don't really understand why that is even happening, and it doesn't happen consistently enough to reliably debug.

Sjors commented:
I also got a not reproducible failure at 65b2da99fe18768d1bd4cd67522f7031f16a3420:

```
228/283 - wallet_migration.py failed, Duration: 2 s

stdout:
2023-11-14T06:24:21.154000Z TestFramework (INFO): PRNG seed is: 2952483070848634871
...
2023-11-14T06:24:22.297000Z TestFramework (INFO): Test migration of an encrypted wallet
2023-11-14T06:24:22.892000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
...
  File "/home/sjors/dev/bitcoin/test/functional/wallet_migration.py", line 452, in test_encrypted
    assert_raises_rpc_error(-4, "The passphrase contains a null character", wallet.migratewallet, None, "pass\0with\0null")
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/util.py", line 130, in assert_raises_rpc_error
    assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/util.py", line 150, in try_rpc
    raise AssertionError("Unexpected exception raised: " + type(e).__name__)
AssertionError: Unexpected exception raised: RemoteDisconnected
2023-11-14T06:24:22.944000Z TestFramework (INFO): Stopping nodes
[node 0] Cleaning up leftover process
```

Sjors commented:
While trying to test the first commit, I ran into something odd. It seems that an _empty_ watch-only wallet legacy is treated as a descriptor wallet by `migratewallet`.

1. Checkout master at 5800c558eb5efb4839ed00d6967e43306d68e1c3, compile with BDB
2. `src/bitcoind -deprecatedrpc=create_bdb`
3. `src/bitcoin-cli createwallet TestMigrate true true "" false false` (this warns that legacy wallets are deprecated)
4. Restart the node (not sure if necessary) and load the wallet again, double check with `getwalletinfo` that it's a legacy wallet
5. `src/bitcoin-cli -rpcwallet=TestMigrate migratewallet`

That results in:

```
error code: -4
error message:
Error: This wallet is already a descriptor wallet
```

Once I import an address it does migrate.

Sjors commented:
Other than that a47f220933f202dd1f972d30f6cb31b09ca0a225 works. Without it, when migrating a watch-only wallet it rescans from the wallet birth height. With it, it does not rescan.

Regarding 65b2da99fe18768d1bd4cd67522f7031f16a3420 this gets a bit unwieldy because there are many places the migration can fail. Maybe just let the caller `MigrateLegacyToDescriptor` attempt a reload if migration failed.

fanquake commented:
Is (some part of) this meant for backport to 26.x?

furszy reviewed (commented):
Nice findings 👌🏼.

I haven't gone deeper over the BDB freeing issue within this context yet. But, can report that in the branch where I changed the underlying db migration mechanism ([link](https://github.com/furszy/bitcoin-core/tree/2023_wallet_safer_migration)), I haven't encountered the issue. I just pulled and ran all this PR new tests there and they all pass successfully in a loop.

One of the ideas of this work is to fix failures like 65b2da9 and others in a safer manner. But can find the detailed reasoning behind it here https://github.com/bitcoin/bitcoin/pull/28609#discussion_r1367739048.

Note: the all-in-one commit in my branch is currently quite messy. The priority was to validate the feasibility of the approach first. Ensuring all tests pass before making any code cleanup and atomic commits reorg. 

Will try to submit a PR with the changes next week.

achow101 commented:
> While trying to test the first commit, I ran into something odd. It seems that an _empty_ watch-only wallet legacy is treated as a descriptor wallet by `migratewallet`.

This is a separate issue in master. I've opened #28976 to address it.

achow101 commented:
Figured out the random failure, should be all working now.

furszy reviewed (commented):
ha, made the same change two hours ago. https://github.com/furszy/bitcoin-core/commit/193e0c9c7ffe34ab2e8a4f5b95f66ecc4bac2791. Will review.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3922,30 +3922,32 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
-                    watchonly_batch->WriteTx(data.watchonly_wallet->mapWallet.at(hash));
-                    // Mark as to remove from this wallet
+        // Check it is the watchonly wallet's
+        // solvable_wallet doesn't need to be checked because transactions for those scripts weren't being watched for
+        if (data.watchonly_wallet) {
+            LOCK(data.watchonly_wallet->cs_wallet);
+            if (data.watchonly_wallet->IsMine(*wtx->tx) || data.watchonly_wallet->IsFromMe(*wtx->tx)) {
+                // Add to watchonly wallet
+                const uint256& hash = wtx->GetHash();
+                const CWalletTx& to_copy_wtx = *wtx;
+                if (!data.watchonly_wallet->LoadToWallet(hash, [&](CWalletTx& ins_wtx, bool new_tx) EXCLUSIVE_LOCKS_REQUIRED(data.watchonly_wallet->cs_wallet) {
+                    if (!new_tx) return false;
+                    ins_wtx.SetTx(to_copy_wtx.tx);
+                    ins_wtx.CopyFrom(to_copy_wtx);
+                    return true;
+                })) {
+                    error = strprintf(_("Error: Could not add watchonly tx %s to watchonly wallet"), wtx->GetHash().GetHex());
+                    return false;
+                }
+                watchonly_batch->WriteTx(data.watchonly_wallet->mapWallet.at(hash));
+                // Mark as to remove from the migrated wallet only if it does not also belong to it
+                if (!IsMine(*wtx->tx) && !IsFromMe(*wtx->tx)) {
                     txids_to_delete.push_back(hash);
-                    continue;
                 }
+                continue;
             }
+        }
+        if (!IsMine(*wtx->tx) && !IsFromMe(*wtx->tx)) {
```
tiny nit:
Could save the "is mine" result at the top and use it in both if/else paths.
```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
--- a/src/wallet/wallet.cpp	(revision ee14243fa526cfaa0ba538286588bb27e13d8c3c)
+++ b/src/wallet/wallet.cpp	(revision e13fcf39a74a45696d1536017b13e5dd393a8ae3)
@@ -3922,31 +3922,32 @@
         }
     }
     for (const auto& [_pos, wtx] : wtxOrdered) {
-        if (!IsMine(*wtx->tx) && !IsFromMe(*wtx->tx)) {
-            // Check it is the watchonly wallet's
-            // solvable_wallet doesn't need to be checked because transactions for those scripts weren't being watched for
-            if (data.watchonly_wallet) {
-                LOCK(data.watchonly_wallet->cs_wallet);
-                if (data.watchonly_wallet->IsMine(*wtx->tx) || data.watchonly_wallet->IsFromMe(*wtx->tx)) {
-                    // Add to watchonly wallet
-                    const uint256& hash = wtx->GetHash();
-                    const CWalletTx& to_copy_wtx = *wtx;
-                    if (!data.watchonly_wallet->LoadToWallet(hash, [&](CWalletTx& ins_wtx, bool new_tx) EXCLUSIVE_LOCKS_REQUIRED(data.watchonly_wallet->cs_wallet) {
-                        if (!new_tx) return false;
-                        ins_wtx.SetTx(to_copy_wtx.tx);
-                        ins_wtx.CopyFrom(to_copy_wtx);
-                        return true;
-                    })) {
-                        error = strprintf(_("Error: Could not add watchonly tx %s to watchonly wallet"), wtx->GetHash().GetHex());
-                        return false;
-                    }
-                    watchonly_batch->WriteTx(data.watchonly_wallet->mapWallet.at(hash));
-                    // Mark as to remove from this wallet
-                    txids_to_delete.push_back(hash);
-                    continue;
-                }
-            }
-            // Both not ours and not in the watchonly wallet
+        bool is_mine = IsMine(*wtx->tx) || IsFromMe(*wtx->tx);
+        // Check it is the watchonly wallet's
+        // solvable_wallet doesn't need to be checked because transactions for those scripts weren't being watched for
+        if (data.watchonly_wallet) {
+            LOCK(data.watchonly_wallet->cs_wallet);
+            if (data.watchonly_wallet->IsMine(*wtx->tx) || data.watchonly_wallet->IsFromMe(*wtx->tx)) {
+                // Add to watchonly wallet
+                const uint256& hash = wtx->GetHash();
+                const CWalletTx& to_copy_wtx = *wtx;
+                if (!data.watchonly_wallet->LoadToWallet(hash, [&](CWalletTx& ins_wtx, bool new_tx) EXCLUSIVE_LOCKS_REQUIRED(data.watchonly_wallet->cs_wallet) {
+                    if (!new_tx) return false;
+                    ins_wtx.SetTx(to_copy_wtx.tx);
+                    ins_wtx.CopyFrom(to_copy_wtx);
+                    return true;
+                })) {
+                    error = strprintf(_("Error: Could not add watchonly tx %s to watchonly wallet"), wtx->GetHash().GetHex());
+                    return false;
+                }
+                watchonly_batch->WriteTx(data.watchonly_wallet->mapWallet.at(hash));
+                // Mark as to remove from this wallet
+                if (!is_mine) txids_to_delete.push_back(hash);
+                continue;
+            }
+        }
+        // Both not ours and not in the watchonly wallet
+        if (!is_mine) {
             error = strprintf(_("Error: Transaction %s in wallet cannot be identified to belong to migrated wallets"), wtx->GetHash().GetHex());
             return false;
         }
```

achow101 replied:
Done as suggested

furszy reviewed (commented):
Code review ACK 0cf1ae5a

achow101 committed:
wallet: Write bestblock to watchonly and solvable wallets

When migrating, we should also be writing the bestblock record to the
watchonly and solvable wallets to avoid rescanning on the reload as that
can be slow.

achow101 committed:
wallet: Reload the wallet if migration exited early

Migration will unload loaded wallets prior to beginning. It will then
perform some checks which may exit early. Such unloaded wallets should
be reloaded prior to exiting.

achow101 committed:
wallet: Keep txs that belong to both watchonly and migrated wallets

It is possible for a transaction that has an output that belongs to the
mgirated wallet, and another output that belongs to the watchonly
wallet. Such transaction should appear in both wallets during migration.

achow101 committed:
test: Test migration of tx with both spendable and watchonly

achow101 committed:
test: Make sure that migration test does not rescan on reloading

We want to make sure that all of the transactions are being copied to
the watchonly and solvable wallets as expected. The automatic rescanning
behavior can cause us to pass a test by finding the transaction
on loading rather than having it be copied as expected.

furszy reviewed (approved):
reACK b093e5e

Only cleaned the IsMine && IsFromMe code dup from my last review.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4210,6 +4212,21 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(const std::string& walle
+    // both before and after reloading. This ensures the set is complete even if one of the wallets
+    // fails to reload.
+    std::set<fs::path> wallet_dirs;
+    // Helper to reload as normal for some of our exit scenarios
```
In commit "wallet: Reload the wallet if migration exited early" (d8002ed333e13e54726bcd922303cb1e42544b44)

I don't think it is good to move `reload_wallet` here before `BackupWallet` is called. It seems like it would make it easy to add an innocent call to `reload_wallet(local_wallet)` that looks like it is just reloading the wallet, but doesn't return right away, so the wallet get added to `wallet_dirs` before it is backed up and possibly deleted as part of the `fs::remove_all` loop.

Most of the code in `reload_wallet` is also unnecessary for handling early returns, in addition to being potentially unsafe. I think the simplest way to handle early returns would just be to add a new simpler lambda, or maybe even inline the LoadWallet call like

```c++
if (was_loaded) {
    local_wallet = LoadWallet(context, local_wallet->GetName(), /*load_on_start=*/std::nullopt, options, status, error, warnings);
}
```

achow101 replied:
I've removed the `wallet_dirs.insert` from `reload_wallet` and moved that to above the three `reload_wallet` calls that actually need it. The lambda is still useful since we have to unload `local_wallet` every time.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4191,11 +4191,13 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(const std::string& walle
     std::vector<bilingual_str> warnings;
 
     // If the wallet is still loaded, unload it so that nothing else tries to use it while we're changing it
+    bool was_loaded = false;
```
In commit "wallet: Reload the wallet if migration exited early" (d8002ed333e13e54726bcd922303cb1e42544b44)

I was confused by "As these checks do not change anything in the wallet, such unloaded wallets should be reloaded prior to exiting" in the commit description, since the wallets would need to be reloaded whether or not checks changed anything.

I think the commit message can just say "Such unloaded wallets should be reloaded prior to exiting" without the "As these checks" part

achow101 replied:
Updated the commit message.

ryanofsky reviewed (commented):
Code review ACK b093e5ea29efa44626db8aeff05ad2ee3296e707, but only lightly reviewed the tests.

This PR has a merge conflict currently so needs to be updated. Also I think the place where reload_wallet lamba is moved is a little unsafe so would suggest changing that (see below).

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4286,24 +4306,19 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(const std::string& walle
-            return to_reload != nullptr;
-        };
         // Reload the main wallet
+        wallet_dirs.insert(fs::PathFromString(local_wallet->GetDatabase().Filename()).parent_path());
```
In commit "wallet: Reload the wallet if migration exited early" (78ba0e6748d2a519a96c41dea851e7c43b82f251)

Not related to this PR, but I think the `fs::PathFromString(wallet->GetDatabase().Filename()).parent_path()` expression repeated throughout this function is unnecessarily verbose and also potentially dangerous if it is used in another context, or if the surrounding context changes.

 It happens to be safe currently, because we know this code is only called after MigrateToSQLite is called, so the database `FileName()` method will always return the path to a file in a wallet subdirectory. But in other contexts `wallet->GetDatabase().Filename()` could return the path to a file in a top level wallets directory, or even a top level data directory, so calling `fs::remove_all(fs::PathFromString(wallet->GetDatabase().Filename()).parent_path())` could wipe out a lot of other data not part of the wallet we are trying to delete.

To fix this as a start, I think we should be deleting the original wallet path
`AbsPathJoin(GetWalletDir(), fs::PathFromString(wallet->GetName()))`, not the parent of the path of the database inside the wallet. It would also be good to move away from using `fs::remove_all` which could delete a lot of files unintentially, and instead write a wallet delete function that only deletes files we know about.

ryanofsky reviewed (approved):
Code review ACK 4da76ca24725eb9ba8122317e04a6e1ee14ac846. This looks great. The code is actually cleaner than before, two bugs are fixed, and the test checking for rescanning is pretty clever and broadens test coverage.

I left a review comment about improving derivation of wallet paths in the MigrateLegacyToDescriptor function, but that isn't really about this PR and would be better addressed in a separate followup.

ryanofsky commented:
If @furszy wants to re-ack, or if someone else adds a review, I think this could be merged

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -4238,8 +4240,20 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(const std::string& walle
+    };
+
     // Before anything else, check if there is something to migrate.
     if (local_wallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
+        if (was_loaded) {
+            reload_wallet(local_wallet);
+        }
         return util::Error{_("Error: This wallet is already a descriptor wallet")};
```
in 78ba0e6748d:

I know that this shouldn't fail, but it would be good to log the error if it happens and inform the user in the error message that they need to reload the wallet manually.

(same for the others)

furszy reviewed (commented):
Code review ACK 4da76ca2

Also verified that the tests fail when bug fix commits are reverted.
